lua/bufferline/jump_mode.lua

```diff
@@ -20,36 +20,41 @@ local state = require'bufferline.state'
 -- Section: Buffer-picking mode state --
 ----------------------------------------
 
-local M = {
+--- @class bufferline.JumpMode
+--- @field private buffer_letter_bimap table<integer|string, integer|string> a bi-directional map of buffer integers and their letters.
+--- @field private index_by_letter table<string, integer> `letters` in the order they were provided
+--- @field private letter_status table<integer, boolean>
+--- @field private letters string the letters to allow while jumping
+--- @field private reinitialize boolean whether to run `initialize_indexes` on next jump.
+local JumpMode = {
   letters = vim.g.bufferline.letters, -- array
   index_by_letter = {}, -- object
   letter_status = {}, -- array
-  buffer_by_letter = {}, -- object
-  letter_by_buffer = {}, -- object
+  buffer_letter_bimap = {}, -- object
```

Comment:
This was a performance improvement. Previously, we had two maps:

1. `buffer_by_letter`, a `table<string, integer>` which is keyed on letter.
2. `letter_by_buffer`, a `table<integer, string>` which is keyed on buffer number.

However, since the (buffer number, letter) pairing forms a [bijection](https://en.wikipedia.org/wiki/Bijection), of different data types, we can instead use a bidirectional map here. Instead of using `buffer_by_letter[letter]`, we can use `buffer_letter_bimap[letter]` (same for `letter_by_buffer`) and get rid of an entire extra `table` in memory.

Reply:
I don't like it. First, because the memory cost is probably around 2x64 bytes for a lua table in itself (the key/values are there one way or another). Second, because the typings become unclear; you can't know for sure what you'll get out (letter or index).